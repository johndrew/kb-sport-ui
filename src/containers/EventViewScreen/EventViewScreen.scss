@import '../../common.scss';

.eventViewScreen__container {
    display: grid;
    grid-template-columns: 10% auto;
}

.eventViewScreen__eventsIcon {
    position: absolute;
    top: 10px;
    left: 10px;
    width: 100px;
    height: 100px;
    h2 {
        font-size: $iconMiniLabel;
    }
}

.eventViewScreen__eventType,
.eventViewScreen__eventDuration {
    color: $tan;
}