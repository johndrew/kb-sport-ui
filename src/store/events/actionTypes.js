export const RANKING_RETRIEVED = 'events.RANKING_RETRIEVED';