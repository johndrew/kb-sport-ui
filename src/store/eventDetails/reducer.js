import { fromJS } from 'immutable';
import * as types from './actionTypes';

export const initialState = fromJS({
    details: [],
});

export default function reduce(state = initialState, action = {}) {

    switch (action.type) {
        case types.DETAILS_FETCHED:
            return state.merge({
                details: action.details,
            });
        default:
            return state;
    }
}

/* SELECTORS */

export function getAllEventDetails(state) {

    return state.eventDetails.get('details');
}

export function getDetailsForEvent(state, eventId) {

    return getAllEventDetails(state).filter(detail => detail.get('eventId') === eventId);
}

export function getRegisteredLifters(state, eventId) {

    return getDetailsForEvent(state, eventId).map(event => event.get('lifterId'));
}

/**
 * Determines if a lifter is registered to any event
 * @returns {Boolean}
 */
export function isLifterRegistered(details, lifterId) {

    return details.find(detail => detail.get('lifterId') === lifterId) != undefined;
}

export function getNumberOfRegisteredEvents(details, lifterId) {

    return details.reduce((total, detail) => (detail.get('lifterId') === lifterId ? total + 1 : total), 0);
}