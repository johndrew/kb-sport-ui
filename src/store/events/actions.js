import * as types from './actionTypes';

export function fetchEvents() {

    return {
        type: types.FETCH_EVENTS,
    };
}