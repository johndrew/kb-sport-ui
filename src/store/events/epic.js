import { ajax } from 'rxjs/ajax';
import { map, mergeMap, switchMap } from 'rxjs/operators';
import { timer } from 'rxjs';
import { ofType } from 'redux-observable';
import * as types from './actionTypes';

const URL = 'https://nfg0lrs4g1.execute-api.us-west-2.amazonaws.com/dev/events';

// action creators
const fetchEventsFulfilled = events => ({ type: types.EVENTS_FETCHED, events });

// epic
export function fetchEvents(action$, state$) {

  return action$.pipe(
    ofType(types.FETCH_EVENTS),
    switchMap(params => {
      return timer(0, 1000).pipe(
        mergeMap(action =>
          ajax.getJSON(URL).pipe(
            map(response => fetchEventsFulfilled(response))
          )
        )
      )
    })
  );
}