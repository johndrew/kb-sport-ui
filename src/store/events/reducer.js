import { fromJS } from 'immutable';
import * as types from './actionTypes';

const initialState = fromJS({
  tmp: {
    ranking: null,
  },
});

export default function reduce(state = initialState, action = {}) {
  switch (action.type) {
    case types.RANKING_RETRIEVED:
      return state.merge({
        tmp: {
          ranking: action.ranking,
        },
      });
    default:
      return state;
  }
}

// SELECTORS

export function getRanking(state) {
  return state.events.getIn(['tmp', 'ranking']);
}