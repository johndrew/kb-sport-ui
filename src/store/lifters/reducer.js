import { fromJS } from 'immutable';
import * as types from './actionTypes';

export const initialState = fromJS({
    lifters: [],
});

export default function reduce(state = initialState, action = {}) {

    switch (action.type) {
        case types.LIFTERS_FETCHED:
            return state.merge({
                lifters: action.lifters,
            });
        default:
            return state;
    }
}

// SELECTORS

export function getAllLifters(state) {

    return state.lifters.get('lifters');
}

export function getLifters(state, lifterIds) {

    return getAllLifters(state).filter(lifter => lifterIds.includes(lifter.get('lifterId')));
}

export function getLiftersBesides(state, lifterIds) {

    return getAllLifters(state).filter(lifter => !lifterIds.includes(lifter.get('lifterId')));
}

export function filterByGender(lifters, gender) {

    return gender === 'female' ?
        lifters.filter(lifter => lifter.get('gender') !== 'women') : 
        lifters.filter(lifter => lifter.get('gender') !== 'men');     
}
