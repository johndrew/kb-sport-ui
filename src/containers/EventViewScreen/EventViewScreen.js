import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import * as eventSelectors from '../../store/events/reducer';
import * as liftersSelectors from '../../store/lifters/reducer';
import * as eventDetailsSelectors from '../../store/eventDetails/reducer';
import * as routes from '../../routes';
import * as eventActions from '../../store/events/actions';
import * as lifterActions from '../../store/lifters/actions';
import * as eventDetailsActions from '../../store/eventDetails/actions';
import EventsIcon from '../../components/presentational/EventsIcon/EventsIcon';
import AddBox from '../../components/presentational/AddBox/AddBox';
import LifterBox from '../../components/presentational/LifterBox/LifterBox';
import BoxCollection from '../../components/wrappers/BoxCollection/BoxCollection';
import ModalWrapper from '../../components/wrappers/ModalWrapper/ModalWrapper';
import ChooseLifter from '../../components/forms/ChooseLifter/ChooseLifter';
import MaleFemaleToggle from '../../components/controls/MaleFemaleToggle/MaleFemaleToggle';
import './EventViewScreen.scss';

class EventViewScreen extends Component {

    constructor(props) {

        super(props);

        this.state = {
            showFemaleLifters: true,
        }

        this.handleGenderToggle = this.handleGenderToggle.bind(this);
    }

    render() {

        if (!this.props.event) return <p>Loading...</p>;

        const liftersToDisplay = this.state.showFemaleLifters ? this.props.registeredFemaleLifters : this.props.registeredMaleLifters;
        const lifters = liftersToDisplay
        .sortBy(lifter => lifter.get('firstName'))
        .map(lifter =>
            <LifterBox
                key={lifter.get('lifterId')}
                lifter={lifter}
                event={this.props.event}
                eventDetails={
                    this.props.eventDetails.find(detail =>
                        detail.get('lifterId') === lifter.get('lifterId') &&
                        detail.get('eventId') === this.props.event.get('eventId')
                    )
                }
                enableModal={true}
                eventView={true} />
        );

        return (
            <div className="eventViewScreen__container">
                <aside className="eventViewScreen__aside">
                    <Link to={routes.EVENTS}>
                        <EventsIcon className="eventViewScreen__eventsIcon" />
                    </Link>
                    <h2 className="eventViewScreen__eventType">
                        {this.props.event.get('type')}
                    </h2>
                    <h2 className="eventViewScreen__eventDuration">
                        {this.props.event.get('duration')}
                    </h2>
                    <MaleFemaleToggle toggled={this.handleGenderToggle}/>
                </aside>
                <div className="eventViewScreen__liftersContainer">
                    <BoxCollection>
                        {lifters}
                        <ModalWrapper
                            triggerComponent={({ open }) => <AddBox addClicked={open}/>}>
                            {({ close }) =>
                                <ChooseLifter
                                    eventId={this.props.event.get('eventId')}
                                    gender={this.state.showFemaleLifters ? 'female' : 'male'}
                                    lifterChosen={close} />
                            }
                        </ModalWrapper>
                    </BoxCollection>
                </div>
            </div>
        );
    }

    componentDidMount() {
  
      this.props.dispatch(eventActions.fetchEvents());
      this.props.dispatch(lifterActions.fetchLifters());
      this.props.dispatch(eventDetailsActions.fetchDetails());
    }

    handleGenderToggle() {

        this.setState({
            showFemaleLifters: !this.state.showFemaleLifters,
        });
    }
}

function mapStateToProps(state, ownProps) {

    const event = eventSelectors.getEvent(state, ownProps.eventId);
    const eventDetails = eventDetailsSelectors.getDetailsForEvent(state, ownProps.eventId);
    const registeredLifters = liftersSelectors.getLifters(state, eventDetailsSelectors.getRegisteredLifters(state, ownProps.eventId));

    return {
        event,
        eventDetails,
        registeredLifters,
        registeredFemaleLifters: liftersSelectors.filterByGender(registeredLifters, 'female'),
        registeredMaleLifters: liftersSelectors.filterByGender(registeredLifters, 'male'),
    };
}

export default connect(mapStateToProps)(EventViewScreen);
