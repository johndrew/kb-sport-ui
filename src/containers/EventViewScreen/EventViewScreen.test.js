import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import EventViewScreen from './EventViewScreen';
import { initialState as initialEventState } from '../../store/events/reducer';
import { initialState as initialLifterState } from '../../store/lifters/reducer';

describe(__filename, () => {

    let mockStore;
    beforeAll(() => {
        
        mockStore = configureStore([]);
    });

    it('renders correctly', () => {

        const wrapper = shallow(
            <EventViewScreen
                store={mockStore({
                    events: initialEventState,
                    lifters: initialLifterState,
                })}
                eventId='foo' />
        );

        expect(wrapper).toMatchSnapshot();
    });
});