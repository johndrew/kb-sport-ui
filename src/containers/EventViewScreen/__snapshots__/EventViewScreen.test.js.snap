// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/Users/taylorjohndrew/Dev/projects/kb-sport-ui/src/containers/EventViewScreen/EventViewScreen.test.js renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(EventViewScreen)
    eventId="foo"
    store={
        Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "dispatch": [Function],
      "event": undefined,
      "eventId": "foo",
      "registeredFemaleLifters": Immutable.List [
],
      "registeredLifters": Immutable.List [
],
      "registeredMaleLifters": Immutable.List [
],
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "dispatch": [Function],
        "event": undefined,
        "eventId": "foo",
        "registeredFemaleLifters": Immutable.List [
],
        "registeredLifters": Immutable.List [
],
        "registeredMaleLifters": Immutable.List [
],
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
